cpp/src_prims/sparse/distance.cuh

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cudart_utils.h>
+#include <sparse/utils.h>
+#include <common/device_buffer.hpp>
+#include <cuml/common/cuml_allocator.hpp>
+#include <raft/cuda_utils.cuh>
+#include <sparse/csr.cuh>
+
+#include <cuml/distance/distance_type.h>
+#include <cuml/neighbors/knn.hpp>
+
+#include <cusparse_v2.h>
+#include <raft/sparse/cusparse_wrappers.h>
+
+namespace MLCommon {
+namespace Sparse {
+namespace Distance {
+
+template <typename value_idx, typename value_t>
+struct distances_config_t {
+  // left side
+  value_idx a_nrows;
+  value_idx a_ncols;
+  value_idx a_nnz;
+  value_idx *a_indptr;
+  value_idx *a_indices;
+  value_t *a_data;
+
+  // right side
+  value_idx b_nrows;
+  value_idx b_ncols;
+  value_idx b_nnz;
+  value_idx *b_indptr;
+  value_idx *b_indices;
+  value_t *b_data;
+
+  cusparseHandle_t handle;
+
+  std::shared_ptr<deviceAllocator> allocator;
+  cudaStream_t stream;
+};
+
+template <typename value_t>
+class distances_t {
+ public:
+  virtual void compute(value_t *out) { CUML_LOG_DEBUG("INside base"); }
+  virtual ~distances_t() = default;
+};
+
+/**
+ * Simple inner product distance with sparse matrix multiply
+ */
+template <typename value_idx = int, typename value_t = float>
+class ip_distances_t : public distances_t<value_t> {
+ public:
+  /**
+   * Computes simple sparse inner product distances as sum(x_y * y_k)
+   * @param[in] config specifies inputs, outputs, and sizes
+   */
+  explicit ip_distances_t(distances_config_t<value_idx, value_t> config)
+    : config_(config),
+      workspace(config.allocator, config.stream, 0),
+      csc_indptr(config.allocator, config.stream, 0),
+      csc_indices(config.allocator, config.stream, 0),
+      csc_data(config.allocator, config.stream, 0),
+      alpha(1.0) {
+    init_mat_descriptor(matA);
+    init_mat_descriptor(matB);
+    init_mat_descriptor(matC);
+    init_mat_descriptor(matD);
+
+    CUSPARSE_CHECK(cusparseCreateCsrgemm2Info(&info));
+
+    CUSPARSE_CHECK(
+      cusparseSetPointerMode(config.handle, CUSPARSE_POINTER_MODE_HOST));
```

Comment:
pointer mode host is the default. This is just a safety net in case it's changed. Are you thinking we should store off the pointer mode that's given as input and change it back when done or that we should just always guarantee that it's set to its default?

Reply:
yes Corey. I was thinking of doing a `cusparseGetPointerMode` before this call and then during the destructor a `cusparseSetPointerMode` to this value.

Because when someone who's relying on mode=device will be in for a surprise if they end up calling this primitive.